base/ltfinal.dtx

```diff
@@ -242,7 +242,70 @@
 % \end{macro}
 % \end{macro}
 %
-%
+% \begin{macro}{trace_stack_levels}
+%   Now define the Lua function to emulate \cs{tracingstacklevels} and
+%   install it in the \texttt{input_level_string} callback.
+% \changes{v2.0k}{2016/01/05}{Remove duplicated code}
+%    \begin{macrocode}
+%</2ekernel>
+%<*2ekernel|latexrelease>
+%<latexrelease>\IncludeInRelease{2021/06/01}{trace_stack_levels}%
+%<latexrelease>                 {Lua trace_stack_levels function}%
+\ifx\directlua\@undefined
+\else
+%<*2ekernel>
+  \expanded{%
+    \everyjob{\the\everyjob
+    \noexpand%\directlua
+%</2ekernel>
+    \directlua{%
+      local function trace_stack_levels (input_ptr)
+        local tracingstacklevels = tex.count.tracingstacklevels
+        if tex.tracingmacros > 0 or input_ptr < tracingstacklevels then
+          if tracingstacklevels > 0 then
+            if input_ptr < tracingstacklevels then
+              return "\string\n\string~" .. string.rep(".",input_ptr)
+            else
+              return "\string~\string~"
+            end
+          else
+            return "\string\n"
+          end
+        else
+          return ""
+        end
+      end
+%<*latexrelease>
+      if luatexbase.callbacktypes['input_level_string'] and %
+         not luatexbase.in_callback('input_level_string','tracingstacklevels') then
+%</latexrelease>
+      luatexbase.add_to_callback('input_level_string',
+        trace_stack_levels,'tracingstacklevels')
+%<*latexrelease>
+      end
+%</latexrelease>
+    }%
+%<*2ekernel>
+  }}%
+%</2ekernel>
+\fi
+%<latexrelease>\EndIncludeInRelease
+%<latexrelease>
+%<latexrelease>\IncludeInRelease{0000/00/00}{trace_stack_levels}%
+%<latexrelease>                 {Lua trace_stack_levels function}%
+%<latexrelease>\ifx\directlua\@undefined
+%<latexrelease>\else
+%<latexrelease>  \directlua{%
+%<latexrelease>    if luatexbase.callbacktypes['input_level_string'] and %
+%<latexrelease>       luatexbase.in_callback('input_level_string','tracingstacklevels') then
+%<latexrelease>        luatexbase.remove_from_callback('input_level_string','tracingstacklevels')
+%<latexrelease>    end}%
```

Comment:
Maybe run this outside of the `\IncludeInRelease` block in order to always redefine the callback? Then we don't get issues later if we ever change the function and `latexrelease` doesn't pick up the changes.

Reply:
Good point.  Done in b26e7ed7